File: microservices/options/options/Models/SupplierOption.js
const mongoose = require("mongoose");
const slugify = require("slugify");
const Schema = mongoose.Schema;
const ObjectId = Schema.Types.ObjectId;

const SupplierOptionSchema = Schema({
    sort: { type: Number, required: false, default: 0 },
    tenant_name: { type: String, required: true },
    tenant_id: { type: String, required: true },
    name: { type: String, required: true, unique_with: 'tenant_id' },
    display_name: { type: Array, required: true },
    source_slug: { type: Array, required: false, default: [] },
    slug: { type: String, required: false, unique_with: 'tenant_id' },
    system_key: { type: String, default: '' },
    description: { type: String, required: false, default: '' },
    information: { type: String, required: false, default: ''},
    media: { type: Array, default: [] },
    incremental_by: { type: Number, required: false },

    published: { type: Boolean, default: true },
    has_children: { type: Boolean, default: false },
    input_type: { type: String, required: false, default: 'radio' },
    extended_fields: { type: Array, default: [] },

    linked: { type: ObjectId, ref: "options", required: false, index: true, default: null },
    shareable: { type: Boolean, required: false, default: false },

    sku: { type: String, default: '', unique: false },
    dimension: { type: String, default: '2d' },
    dynamic: { type: Boolean, default: false },
    unit: { type: String, required: false, default: 'mm' },
    width: { type: Number, default: 0 },
    maximum_width: { type: Number, default: 0 },
    minimum_width: { type: Number, default: 0 },
    height: { type: Number, default: 0 },
    maximum_height: { type: Number, default: 0 },
    minimum_height: { type: Number, default: 0 },
    length: { type: Number, default: 0 },
    maximum_length: { type: Number, default: 0 },
    minimum_length: { type: Number, default: 0 },
    parent: { type: Boolean, default: true },
    start_cost: { type: Number, default: 0 },
    rpm: { type: Number, default: 0 },

    sheet_runs: [
        {
            _id: false,
            machine: ObjectId,
            dlv_production: { type: Array, default: [] },
            runs: { type: Array, default: [] }
        }
    ],
    runs: [
        {
            _id: false,
            category_id: ObjectId,
            start_cost: { type: Number, default: 0 },
            runs: { type: Array, default: [] }
        }
    ],
    boxes: { type: Array, default: '' },
    additional: {
        _id: false,
        type: {
            calc_ref: { type: String },
            calc_ref_type: { type: String }
        },
        required: false,
        default: {}
    },
    created_at: { type: Date, default: Date.now },
    calculation_method: { type: String, default: "qty" },
    configure: [
        {
            _id: false,
            category_id: ObjectId,
            configure: {
                incremental_by: { type: Number, default: 0 },
                dimension: { type: String, default: "2d" },
                dynamic: { type: Boolean, default: false },
                dynamic_keys: { type: Array, default: [] },
                start_on: { type: Number, default: 0 },
                end_on: { type: Number, default: 0 },
                generate: { type: Boolean, default: false },
                dynamic_type: { type: String, default: "integer" },
                unit: { type: String, default: "mm" },
                width: { type: Number, default: 0 },
                maximum_width: { type: Number, default: 0 },
                minimum_width: { type: Number, default: 0 },
                height: { type: Number, default: 0 },
                maximum_height: { type: Number, default: 0 },
                minimum_height: { type: Number, default: 0 },
                length: { type: Number, default: 0 },
                maximum_length: { type: Number, default: 0 },
                minimum_length: { type: Number, default: 0 },
                start_cost: { type: Number, default: 0 },
                calculation_method: { type: String, default: "qty" },
                additional: {
                    _id: false,
                    type: {
                        calc_ref: { type: String },
                        calc_ref_type: { type: String }
                    },
                    required: false,
                    default: {}
                }
            }
        }
    ],
    dynamic_keys: { type: Array, required: false, default: [] },
    dynamic_type: { type: String, required: false, default: '' },
    dynamic_object: { type: Array, required: false, default: null },
    end_on: { type: Number, required: false, default: 0 },
    start_on: { type: Number, required: false, default: 0 },
    generate: { type: Boolean, required: false, default: false },
}, {
    versionKey: false
});

SupplierOptionSchema.pre('save', function (next) {
    if (this.name) {
        this.slug = slugify(this.name, { lower: true });
        this.system_key = slugify(this.name, { lower: true });
    }
    next();
});

module.exports = SupplierOption = mongoose.model("supplier_options", SupplierOptionSchema);

File: microservices/assortments/models/supplierOption.py
from database.db import db
from flask_mongoengine import BaseQuerySet
from slugify import slugify, Slugify, UniqueSlugify
from models.option import Option
from models.box import Box
from models.supplierBox import SupplierBox
from mongoengine import ValidationError
import datetime

# def id(): return str(uuid.uuid4())

class SupplierOption(db.Document):
    sort = db.IntField(required=False, default=0)
    tenant_name = db.StringField(required=True)
    tenant_id = db.StringField(required=True)
    name = db.StringField(required=True)
    display_name = db.ListField(required=True)
    slug = db.StringField(required=False, unique_with=('tenant_id'))
    source_slug = db.ListField(required=False)
    system_key = db.StringField(required=False)
    description = db.StringField(required=False, default='')
    information = db.StringField(required=False, default='')
    media = db.ListField(required=False, default=[])
    incremental_by = db.IntField(required=False, default=0)
    published = db.BooleanField(default=True)
    has_children = db.BooleanField(required=False, default=False)
    input_type = db.StringField(required=False, default="radio")
    extended_fields = db.ListField(required=False, default=[])
    linked = db.ReferenceField(Option, required=False,default=None)
    shareable = db.BooleanField(required=False, default=False)
    sku = db.StringField(required=False, default='')
    children = db.ListField(db.ReferenceField('self', default=[]), required=False, default=[])
    parent = db.BooleanField(default=True)
    rpm = db.IntField(required=False, min_value=None, max_value=None, default=0)
    sheet_runs = db.ListField(required=False, default=[])
    runs = db.ListField(required=False, default=[])
    boxes = db.ListField(db.ReferenceField(SupplierBox), required=False, default=[])
    additional = db.DictField(required=False, default={})
    configure = db.ListField(required=False, default=[])
    created_at = db.DateTimeField(required=False, default=datetime.datetime.now)

    dimension = db.StringField(required=False, default='2d')
    dynamic = db.BooleanField(required=False, default=False)
    dynamic_type = db.StringField(required=False, default="")
    dynamic_keys = db.ListField(required=False, default=[])
    start_on = db.IntField(required=False, default=0)
    end_on = db.IntField(required=False, default=0)
    generate = db.BooleanField(required=False, default=False)
    unit = db.StringField(required=False, default="mm")
    width = db.IntField(required=False, min_value=None, max_value=None, default=0)
    maximum_width = db.IntField(required=False, min_value=None, max_value=None, default=0)
    minimum_width = db.IntField(required=False, min_value=None, max_value=None, default=0)
    height = db.IntField(required=False, min_value=None, max_value=None, default=0)
    maximum_height = db.IntField(required=False, min_value=None, max_value=None, default=0)
    minimum_height = db.IntField(required=False, min_value=None, max_value=None, default=0)
    length = db.IntField(required=False, min_value=None, max_value=None, default=0)
    maximum_length = db.IntField(required=False, min_value=None, max_value=None, default=0)
    minimum_length = db.IntField(required=False, min_value=None, max_value=None, default=0)
    start_cost = db.IntField(required=False, default=0)
    calculation_method = db.StringField(required=False, default='qty')

    meta = {
        'indexes': ['-created_at', 'slug', 'name', 'linked'],
        'ordering': ['-created_at', 'name'],
        'collection': 'supplier_options',
        'strict': False,
        'queryset_class': BaseQuerySet
    }

    def __setattr__(self, key, value):
        super(SupplierOption, self).__setattr__(key, value)
        if key == 'name':
            self.slug = slugify(self.name, to_lower=True)
